Jake Vendingmachine step2

ChocOZerO opened the pull request:
자판기 2단계 구현

ChocOZerO committed:
[Add]음료 객체 3계층 이상으로 변경 #1, #2

ChocOZerO committed:
[Add]new interfaces for new properties #3

ChocOZerO committed:
[Add]Test cases for description & method in each #4

ChocOZerO committed:
[Update]음료수 객체 인스턴스를 종류별로 만들어서 출력 #5

ChocOZerO committed:
[FitCodacy]공백정리

godrm commented on VendingMachine/VendingMachine/StrawberryMilk.swift:
```diff
@@ -0,0 +1,22 @@
+    private let taste: String = "딸기우유"
+
+    override init(brand: String, weight: Int, price: Int, name: String, validate: String) {
+        super.init(brand: brand, weight: weight, price: price, name: name, validate: validate)
```
`private let taste: String = "딸기우유"` 이렇게 변수를 초기화하는 것과
init() 생성함수에서 초기화하는 것과 어떤 차이가 있을까요? 어떤 기준에서 나눠서 해야할까요?

ChocOZerO replied:
init안에서 생성하는건 인스턴스를 생성할때마다 입력을 다르게 주어서 생성시점에 변화를 주고 싶은 속성을 집어 넣고, 초기화를 강제하는 경우는 해당 객체로 생성하는 모든 인스턴스에 공통된 속성값을 할당하고 싶을때 사용합니다.

godrm commented on VendingMachine/VendingMachine/Milk.swift:
```diff
@@ -9,22 +9,26 @@
+        dateComponents.year = Int(validate.prefix(4))!
+        dateComponents.month = Int(validate.suffix(4).prefix(2))!
+        dateComponents.day = Int(validate.suffix(2))
+        dateComponents.timeZone = TimeZone(abbreviation: "KST")
```
validate도 formatter 를 이용해서 나눠보세요.

ChocOZerO replied:
String 을 Int로 바꿀때 말씀하시는 건가요? 아니면 DateFormatter를 활용해보라는 건가요?

godrm replied:
String을 이렇게 직접 잘라서 Int로 바꾸는게 아니라 String -> Date로 직접 바꿔보라는겁니다.

ChocOZerO replied:
네 알겠습니다~
한 가지 궁금한게 있는데요, Date같은 경우 테스트 하는 날에 따라 현재시간이 다르게 찍혀서 테스트 케이스가 날짜에 의존적인 경향이 있는데, 이럴 경우 어떤식으로 테스트하는게 좋을까요?

godrm replied:
단위 테스트 코드가 테스트하기 어려운 의존성이 있으면 코드가 좋은 구조가 아니겠죠
그래서 의존성을 객체 생성할 때 외부에서 넣어줄 수 있도록 하는 경우가 많습니다. 
생성하는 Date를 넘겨줄 수 있도록 작성해야 테스트가 가능할 겁니다. 참고하세요.

ChocOZerO committed:
[Update]Milk객체의 Validate를 formatter 활용 방식으로 변경 #6

ChocOZerO committed:
[Refactoring]모든 날짜를 Date타입 주입방식으로 변경 #7

테스트를 용이하게하고 의존성을 줄이기 위해
Date와 String에 extension을 활용해 서로를 얻을 수 있는 메소드 추가
기존: String을 넘겨서 자동 생성
변경: 각 음료 인스턴스 생성 전에 Date인스턴스를 만들어서 주입

ChocOZerO committed:
[Comment]주석 수정

ChocOZerO commented:
Formatter를 활용해서 String을 바로 Date로 변경할 수 있도록 수정.
Date를 활용해서 의존성을 줄이는 방식으로 수정.

godrm commented:
상황에 따라서 꼭 Date 타입이 아닐 수 있지만, 의존성을 주입하는 구조는 익숙해질 필요가 있습니다.
다음 단계를 진행하세요.
